src/rgw/rgw_iam_policy.cc

```diff
@@ -1115,7 +1121,18 @@ Effect Statement::eval(const Environment& e,
   }
   if (!resource.empty() && res) {
     if (!std::any_of(resource.begin(), resource.end(),
-          [&res](const ARN& pattern) {
+          [&res,&e](const ARN& pattern) {
+            if(pattern.resource == "${aws:username}"){
```

Comment:
instead of testing for equality here, i think we need to search the entire resource for this substring and do a string replacement. the [aws doc](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html#policy-vars-using-variables) shows this example:
> "Resource": ["arn:aws:s3:::mybucket/${aws:username}/*"]

so i'd expect `pattern.resource` to contain the string `mybucket/${aws:username}/*`